diff --git a/configure.ac b/configure.ac
index c5b6d6d..4f2b8a0 100644
--- a/configure.ac
+++ b/configure.ac
@@ -9,6 +9,7 @@ AC_CANONICAL_HOST
 AC_ISC_POSIX
 
 AM_INIT_AUTOMAKE(1.6 dist-bzip2)
+AM_MAINTAINER_MODE
 AM_CONFIG_HEADER(config.h)
 
 AC_LIBTOOL_WIN32_DLL
diff --git a/src/lib/Makefile.am b/src/lib/Makefile.am
index 3391569..de86f84 100644
--- a/src/lib/Makefile.am
+++ b/src/lib/Makefile.am
@@ -57,7 +57,7 @@ libeina_la_SOURCES += $(top_srcdir)/src/modules/mp/pass_through/pass_through.c
 endif
 
 libeina_la_LIBADD = @EINA_LIBS@ @dlopen_libs@
-libeina_la_LDFLAGS = -no-undefined @lt_enable_auto_import@ -version-info @version_info@
+libeina_la_LDFLAGS = -no-undefined @lt_enable_auto_import@ -release @PACKAGE_VERSION@$(DEBIAN_MINOR)
 
 clean-local:
 	rm -rf *.gcno
